packages/shared/src/hooks/useUserMention.ts

```diff
@@ -0,0 +1,231 @@
+import { useQuery, useQueryClient } from 'react-query';
+import {
+  useState,
+  useContext,
+  KeyboardEvent as ReactKeyboardEvent,
+  useEffect,
+  MutableRefObject,
+} from 'react';
+import request from 'graphql-request';
+import {
+  RecommendedMentionsData,
+  RECOMMEND_MENTIONS_QUERY,
+} from '../graphql/comments';
+import { apiUrl } from '../lib/config';
+import { UserShortProfile } from '../lib/user';
+import AuthContext from '../contexts/AuthContext';
+import useDebounce from './useDebounce';
+import { isAlphaNumeric, isSpecialCharacter } from '../lib/strings';
+import {
+  CaretPosition,
+  getCaretPostition,
+  getWord,
+  hasSpaceBeforeWord,
+  isBreakLine,
+  replaceWord,
+  setCaretPosition,
+} from '../lib/element';
+import { nextTick } from '../lib/func';
+
+interface UseUserMention {
+  mentionQuery?: string;
+  onMentionKeypress: (event: ReactKeyboardEvent) => unknown;
+  selected: number;
+  mentions: UserShortProfile[];
+  offset: CaretPosition;
+  onInputClick?: () => unknown;
+  onMentionClick?: (username: string) => unknown;
+  onInitializeMention: () => unknown;
+}
+
+interface UseUserMentionProps {
+  postId: string;
+  onInput: (content: string) => unknown;
+  commentRef?: MutableRefObject<HTMLDivElement>;
+}
+
+const ARROW_KEYS = ['ArrowUp', 'ArrowDown'];
+
+export function useUserMention({
```

Comment:
The business logic here of how you handle the different chars like `Backspace` and other long code sections need more documentation I think. It will be hard to understand it in a few days.

Reply:
I have now added more information relating to this on the specs page in the product section at point number 5.